File: client/python/camoucrome/identity.py
"""Per-instance values beside a preset. A preset (settings/presets/*.json) is
a device identity; the seeds are not part of it (preset_loader.cc emits
none), so each launch draws them fresh unless the caller pins them."""
import random

# Mirrors settings/launcher.json "per_instance_seeds"; the test asserts it.
SEED_KEYS = ("canvas:seed", "audio:seed", "mediaDevices:seed")


def per_instance_config(rng=None):
    """Fresh non-zero uint32 seeds for every key in SEED_KEYS. Merge the
    caller's explicit keys over the result; pass it as launch(config=...)
    beside launch(preset=...)."""
    rng = rng or random.SystemRandom()
    return {k: rng.randint(1, 0xFFFFFFFF) for k in SEED_KEYS}

File: client/go/identity.go
package camoucrome

import (
	"crypto/rand"
	"encoding/binary"
)

// SeedKeys mirrors settings/launcher.json "per_instance_seeds". A preset is a
// device identity; these are per instance and never part of it.
var SeedKeys = []string{"canvas:seed", "audio:seed", "mediaDevices:seed"}

// PerInstanceConfig draws a fresh non-zero uint32 for every SeedKeys entry.
// Merge explicit keys over it and pass it as Options.Config beside
// Options.Preset. A nil reader uses crypto/rand.
func PerInstanceConfig(read func([]byte) (int, error)) (map[string]any, error) {
	if read == nil {
		read = rand.Read
	}
	cfg := make(map[string]any, len(SeedKeys))
	for _, k := range SeedKeys {
		var b [4]byte
		for {
			if _, err := read(b[:]); err != nil {
				return nil, err
			}
			if v := binary.LittleEndian.Uint32(b[:]); v != 0 {
				cfg[k] = v
				break
			}
		}
	}
	return cfg, nil
}
